### ISSUE TITLE
[proposal] Move constants and conversion factors in `units.py` to pint

### ISSUE THREAD

**Issue by hmacdope** [MDAnalysis/mdanalysis#3924]
## Is your feature request related to a problem? ##

The units.py file has a large number of hardcoded constants and conversion factors While these are tested and are accurate, I think it might be a bit more useful to define them using `pint` as **relational** constants rather than just hardcoded factors.  This makes adding more units and constants easier and **systematic**

This means something like the way angstrom / fm is defined would move from 
```python
length_unit_factor = {
    'Angstrom': 1.0,
    ''fm': 1e5
```
to  something like 

```python
from pint import UnitRegistry
MDA_PINT_UNITS = UnitRegistry(system='SI')
MDANALYSIS_BASE_PINT_UNITS = {"length": MDA_PINT_UNITS.Unit("angstrom")
length_unit_factor = {
'Angstrom': (1.0*MDANALYSIS_BASE_PINT_UNITS["length"]).to("angstrom").magnitude
'fm': (1.0*MDANALYSIS_BASE_PINT_UNITS["length"]).to("fm").magnitude
}
```

While it is a bit more verbose, it is at least guaranteed to be internally consistent. 

**NB this doesn't change the units API or introduce any `Pint` typing** it only changes how the constants are generated. This is more a soft launch of pint.  We can work that in further down the line if people are interested. 

Additional benefits include 

- Paves the way for further pint inclusion as discussed in #3792 
- **Brings us in line with the redefinition of SI units in 2019 (we are using 2010 definitions)**
-  We no longer have to vendor new units.
- All our units will be self-consistent reducing propensity for errors.
- Allow us to expose optional `pint` quantified things later down the track. 

## Describe the solution you'd like ##

We could move the conversion factors to be much more declarative by using `pint` and expressing the conversion factors as relational
as detailed above 

I have a PR mostly ready to roll if people are interested. @MDAnalysis/coredevs, let me know. :)

## Describe alternatives you've considered ##

leave everything the way it is. 

## Additional context ##

You can move the defintion of units to a .dat file if required as discussed in #3792

**Comment by jbarnoud** [comment]
Years ago we discussed going full on with pint. I do not find the discussion but points I remember are that it was slow and we did not always know them unit.

The approach here seems more conservative, which I like.

Small question, though : if a user already has a pint registry in their program, how does that interact with the registry used by mdanalysis? In a related note, would it be possible to switch to the older unit definition for reproducibility?

**Comment by hmacdope** [comment]
I think it should be fine unless there is already a registry named MDA_PINT_UNITS. Even then it might  have some kind of singleton guard I'm not sure. 

So IMO low chance of cross contamination.

**Comment by jbarnoud** [comment]
That is my understanding as well. It is not an issue as long as we do not expose the types.

**Comment by hmacdope** [comment]
Just for completeness the discussion you mentioned could be #2230 which is basically the same as what I am proposing here.

**Comment by orbeckst** [comment]
I am supportive (as I was for #2230). But we need to ensure that at the point of change-over **no units change values** (see https://github.com/MDAnalysis/mdanalysis/issues/2230#issuecomment-478819179 ).

**Comment by hmacdope** [comment]
@orbeckst  what about changes to values that have been redefined.  Some may be a few decimals different post redefinition.

**Comment by hmacdope** [comment]
A possible idea to avoid change on crossover would be to move the constants defined in the old units.py to be tests?

**Comment by orbeckst** [comment]
That’s a good idea. We initially need to know exactly what would change.

**Comment by hmacdope** [comment]
Let it be known this I am upset with the person that came up with `AKMA` units system

**Comment by hmacdope** [comment]
@MDAnalysis/coredevs I would also like to remove the hardcoded constant water densities eg `mda.units.water` and associated values in `mda.units.densityFactors`.

 I think these are possibly a bit misleading to ship as they provide the density of `SPC` and other water models under **ideal** conditions, rather than the actual value of them in user simulations. This I think is slightly dangerous.

We can still keep the values tablulated, but I would consider refactoring methods that require a water density to take the water density as an input rather than having a hardcoded constant.

Thoughts? This would be a 3.0 target. I can raise a seperate series of issues to move discussion there if required.

**Comment by orbeckst** [comment]
In our typical applications, we need some reference value for water. The literature value at standard conditions for the water model is not the worst choice, in my opinion. Similarly, having the experimental value at hand is also good. Having an option to provide your own density is not bad but for something like membrane simulations not particularly convenient because actually computing what your "bulk" value is, is more effort than it's worth (in my opinion). 

So in summary, I'd like to keep these values around — they have paper references and are established values. Obviously, it's the user's responsibility to worry about temperature and anything else but I think it's sufficient to add a warning box to the docs.

We could move the density values into the Density module, though, especially if this is the only place where they are used. Did you check?

**Comment by orbeckst** [comment]
Would it make sense to make units handling a GSOC project or is @hmacdope 's PR #3941 already too far along to make this worthwhile?

**Comment by hmacdope** [comment]
@orbeckst @MDAnalysis/coredevs could defs do with an assist here.   Lots to do both on PR and in the units space in general I have a few ideas.

**Comment by tylerjereddy** [comment]
Sorry if I'm missing this from another discussion, but from a design standpoint, do we have a justification for preferring `pint` over whatever `astropy` did for [Units and Quantities](https://docs.astropy.org/en/stable/units/index.html)? I thought they might have used something else under the hood?

One other thing to have on the radar, though perhaps less mature, is direct NumPy integration with unit metadata with  [`unyt`](https://unyt.readthedocs.io/en/stable/) and NEP42 per https://github.com/numpy/numpy-user-dtypes/tree/main/unytdtype.

In any case, just want to make sure this is on the radar wrt adopting a new dependency if one of those other things looks more appealing in the long term.
